winter/web/exceptions/problem_handling.py

```diff
@@ -3,65 +3,91 @@
 from typing import Type
 
 import dataclasses
+from dataclasses import asdict
+from dataclasses import dataclass
+from dataclasses import is_dataclass
 from rest_framework.request import Request
 
 from winter.core import Component
 from winter.core.utils import camel_to_human
+from winter.web import ResponseHeader
 from .exception_handler_generator import ExceptionHandlerGenerator
 from .exception_mapper import ExceptionMapper
 from .handlers import ExceptionHandler
 from .handlers import exception_handlers_registry
 from .problem_annotation import ProblemAnnotation
+from .problem_handling_info import ProblemHandlingInfo
+from ... import response_header
 
 
 class ProblemExceptionHandlerGenerator(ExceptionHandlerGenerator):
-    def generate(self, exception_class: Type[Exception], exception_mapper: ExceptionMapper) -> Type[ExceptionHandler]:
-        from winter import response_status
+    def __init__(self, exception_mapper: ExceptionMapper):
+        self._exception_mapper = exception_mapper
 
-        component = Component.get_by_cls(exception_class)
-        annotation = component.annotations.get_one(ProblemAnnotation)
-        assert isinstance(annotation, ProblemAnnotation)
+    def generate(self, exception_class: Type[Exception], handling_info: ProblemHandlingInfo) -> Type[ExceptionHandler]:
+        from winter import response_status
 
         return_type_class = self._build_exception_dataclass(exception_class)
-
-        @response_status(annotation.status)
-        def handle_method(self, request: Request, exception: exception_class, **kwargs) -> return_type_class:
-            return exception_mapper.to_response_body(request, exception)
+        exception_mapper = self._exception_mapper
+
+        @response_status(handling_info.status)
+        @response_header('Content-Type', 'content_type_header')
+        def handle_method(
+            self,
+            request: Request,
+            exception: exception_class,
+            content_type_header: ResponseHeader[str],
+            **kwargs,
+        ) -> return_type_class:
+            content_type_header.set('application/json+problem')
+            return exception_mapper.to_response_body(request, exception, handling_info)
 
         handler_class_name = exception_class.__name__ + 'Handler'
         handler_class = type(handler_class_name, (ExceptionHandler,), {'handle': handle_method})
         return handler_class
 
+    def autodiscover(self):
```

Comment:
`ProblemExceptionHandlerGenerator` is a default implementation, which is optional and could be customized in the future on the used side. Autodiscovery on the other hand is not going to be customizable.
My suggestions:
- Move the method outside of the class.
- Move the method to the base abstract class, and as a consequence allow users to customize the autodiscovery process.
- Your opinion?

Reply:
I don't have a preference. Just extracted it from the class.